File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValidBSTHelper(root, null, null);
    }
    public boolean isValidBSTHelper(TreeNode node, Integer lower, Integer upper) {
        if(node == null)return true;
        int val = node.val;
        if(lower != null && val <= lower)return false;
        if(upper != null && val >= upper)return false;
        if(!isValidBSTHelper(node.right, val, upper))
            return false;
        if(!isValidBSTHelper(node.left, lower, val)) 
            return false;
        
        return true;
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        def helper(root: TreeNode, min_val: float, max_val: float) -> TreeNode:
            if root is None:
                return True

            if not (min_val < root.val < max_val):
                return False

            return helper(root.left, min_val, root.val) and helper(root.right, root.val, max_val)

        return helper(root, float(-inf), float(+inf))